### PR TITLE
Fix perfdata parser not recognize scientific notation

### PR DESCRIPTION
**Pull request by mcktr**
The scientific notation is basically allowed in our performance data
parser. In an edge case where scientific notation is delivered with
an big E letter our parser does not recognize it and drops it as
false performance data.

This wrong interpretation as false performance data affects all parts of
the performance data, the value itself, the warning value, the critical
value, the minimum value and the maximum value.

Since all parts of the performance data are extracted from the valueStr,
it is sufficient to transform it to the lower case.

Example: 
```
[2020-11-19 17:12:36 +0100] warning/InfluxdbWriter: Ignoring invalid perfdata for checkable 'xxx!JMX:Events/1m' and command 'by_ssh' with value: OneMinuteRate=5.863856523964383E-6
```
I could verify the issue locally with the following test script:

```
#!/usr/bin/env bash

echo "OK - Test output | 'OneMinuteRate'=4.1E-1;5.4E-3;10E-5;0;100"
```

https://community.icinga.com/t/influxdbwriter-invalid-perfdata-on-low-values/6109

### CODE REVIEW

File: lib/base/perfdatavalue.cpp [hunk 1]
```diff
@@ -45,6 +45,8 @@ PerfdataValue::Ptr PerfdataValue::Parse(const String& perfdata)
 
 	String valueStr = perfdata.SubStr(eqp + 1, spq - eqp - 1);
 
+	valueStr = valueStr.ToLower();
+
 	size_t pos = valueStr.FindFirstNotOf("+-0123456789.e");
```
Comment (julianbrost) [inline lib/base/perfdatavalue.cpp:50]: Have you tried adding upper-case `E` here? If it doesn't work then, I'd consider this a bug in `Convert::ToDouble`.

Comment (mcktr) [reply lib/base/perfdatavalue.cpp:50]: Adding the upper-case `E` works.
```diff
diff --git a/lib/base/perfdatavalue.cpp b/lib/base/perfdatavalue.cpp
index 6371b54ef..5531dc418 100644
--- a/lib/base/perfdatavalue.cpp
+++ b/lib/base/perfdatavalue.cpp
@@ -45,9 +45,7 @@ PerfdataValue::Ptr PerfdataValue::Parse(const String& perfdata)
 
        String valueStr = perfdata.SubStr(eqp + 1, spq - eqp - 1);
 
-       valueStr = valueStr.ToLower();
-
-       size_t pos = valueStr.FindFirstNotOf("+-0123456789.e");
+       size_t pos = valueStr.FindFirstNotOf("+-0123456789.eE");
 
        double value = Convert::ToDouble(valueStr.SubStr(0, pos));
 
@@ -162,7 +160,7 @@ String PerfdataValue::Format() const
 
 Value PerfdataValue::ParseWarnCritMinMaxToken(const std::vector<String>& tokens, std::vector<String>::size_type index, const String& description)
 {
-       if (tokens.size() > index && tokens[index] != "U" && tokens[index] != "" && tokens[index].FindFirstNotOf("+-0123456789.e") == String::NPos)
+       if (tokens.size() > index && tokens[index] != "U" && tokens[index] != "" && tokens[index].FindFirstNotOf("+-0123456789.eE") == String::NPos)
                return Convert::ToDouble(tokens[index]);
        else {
                if (tokens.size() > index && tokens[index] != "")
```

Which proposal should I implement?

Comment (julianbrost) [reply lib/base/perfdatavalue.cpp:50]: I'd prefer the one that updates the string. Bonus points for making it a constant as it is there in two places.

With the original PR, I guess `tokens[index] != "U"` in the if would always evaluate to true. So maybe another reason to do it the other way? (Note to self/thinking out loud: how would this break and do we want a test-case that covers this?)

**Review by julianbrost — CHANGES_REQUESTED** [review]
Awesome, thanks for writing a test :+1:

To actually run, it also has to be added to `test/CMakeLists.txt`. But apart from that, looking good.

**Review by julianbrost — APPROVED** [review]
:+1:

**Review by Al2Klimov — APPROVED** [review]

**Comment by julianbrost** [comment]
Ok, this breaks in conjunction with #7871. I'll have a look at this tomorrow.

**Pull request merged**